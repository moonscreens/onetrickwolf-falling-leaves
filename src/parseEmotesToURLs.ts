export default function parseEmotesToURLs(emotes: object, message: string): object[] {
  // Things about to get messy...
  const emotesArr:[string, number][] = [];
  // For each emote object break it into its own entry based on the number of positions it has
  // e.g. 304734759":["0-8","20-28"] becomes ['304734759','0-8' ],['304734759','20-28' ]
  Object.entries(emotes).forEach(([emoteId, positions]) => {
    positions.forEach((position:string) => {
      const firstPosition = Number(position.split('-')[0]); // Only need the first position for order
      emotesArr.push([emoteId, firstPosition]);
    });
  });

  emotesArr.sort((a, b) => a[1] - b[1]); // Sort by position
  return emotesArr.map(([emoteId, position]) => {
    const emoteObj = {
      position,
      imageType: 'static',
      url: `https://static-cdn.jtvnw.net/emoticons/v1/${emoteId}/3.0`,
    };
    if (emoteId.includes('emotesv2_')) {
      emoteObj.imageType = 'animated';
      emoteObj.url = `https://static-cdn.jtvnw.net/emoticons/v2/${emoteId}/default/dark/3.0`;
    }
    return emoteObj;
  });
}
