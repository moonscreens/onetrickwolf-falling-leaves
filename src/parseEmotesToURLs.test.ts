import parseEmotesToURLs from './parseEmotesToURLs';

const emoteSample1 = JSON.parse('{"301948071":["0-6"],"301948074":["8-14"]}');
const messageSample1 = 'moon2EE moon2LL';

test('Non-animated Twitch emotes have proper URLs and order', () => {
  expect(parseEmotesToURLs(emoteSample1, messageSample1)).toMatchSnapshot();
});

const emoteSample2 = JSON.parse('{"emotesv2_553fbd45762b4387b204c53285a93dbb":["0-7"],"emotesv2_447df256f3b1412b9fa0dfd3e9b6d84c":["9-17"]}');
const messageSample2 = 'moon2JAM moon2SPIN';

test('Animated Twitch emotes have proper URLs and order', () => {
  expect(parseEmotesToURLs(emoteSample2, messageSample2)).toMatchSnapshot();
});

const emoteSample3 = JSON.parse('{"304734759":["0-8","20-28"],"emotesv2_447df256f3b1412b9fa0dfd3e9b6d84c":["10-18"]}');
const messageSample3 = 'moon2CUTE moon2SPIN moon2CUTE';

test('Duplicate Twitch emotes have proper URLs and order', () => {
  expect(parseEmotesToURLs(emoteSample3, messageSample3)).toMatchSnapshot();
});

// TODO: BTTV emotes
