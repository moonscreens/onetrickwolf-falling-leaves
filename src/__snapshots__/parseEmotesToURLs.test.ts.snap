// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animated Twitch emotes have proper URLs and order 1`] = `
Array [
  Object {
    "imageType": "animated",
    "position": 0,
    "url": "https://static-cdn.jtvnw.net/emoticons/v2/emotesv2_553fbd45762b4387b204c53285a93dbb/default/dark/3.0",
  },
  Object {
    "imageType": "animated",
    "position": 9,
    "url": "https://static-cdn.jtvnw.net/emoticons/v2/emotesv2_447df256f3b1412b9fa0dfd3e9b6d84c/default/dark/3.0",
  },
]
`;

exports[`Duplicate Twitch emotes have proper URLs and order 1`] = `
Array [
  Object {
    "imageType": "static",
    "position": 0,
    "url": "https://static-cdn.jtvnw.net/emoticons/v1/304734759/3.0",
  },
  Object {
    "imageType": "animated",
    "position": 10,
    "url": "https://static-cdn.jtvnw.net/emoticons/v2/emotesv2_447df256f3b1412b9fa0dfd3e9b6d84c/default/dark/3.0",
  },
  Object {
    "imageType": "static",
    "position": 20,
    "url": "https://static-cdn.jtvnw.net/emoticons/v1/304734759/3.0",
  },
]
`;

exports[`Non-animated Twitch emotes have proper URLs and order 1`] = `
Array [
  Object {
    "imageType": "static",
    "position": 0,
    "url": "https://static-cdn.jtvnw.net/emoticons/v1/301948071/3.0",
  },
  Object {
    "imageType": "static",
    "position": 8,
    "url": "https://static-cdn.jtvnw.net/emoticons/v1/301948074/3.0",
  },
]
`;
